Adds a sample of how Crossplane + OAM work with KPT (#31)
oam-dev/samples: 5.OAM_KPT_Demo/README.md

```

Then the app operator could see which parameters are available in this component like below:

```shell
$ kpt cfg list-setters sampleapp/
  NAME              VALUE               SET BY                   DESCRIPTION             COUNT
  image           nginx:1.16.1        sampleapp developer   use to set image for component   1
  instance-name   example-component   sampleapp developer   use to set an instance name      1
```

It's very clear and easy to understand.

#### Set Value by App Operator

Then the application operator could set `instance-name` with a new name like this:

```shell
$ kpt cfg set sampleapp/ instance-name test-component
set 1 fields
```

Check the component and you will find the instane name has been changed.

```shell
$ cat sampleapp/component.yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: test-component # {"$ref":"#/definitions/io.k8s.cli.setters.instance-name"}
spec:
  parameters:
  - name: instance-name
    fieldPaths:
    - metadata.name
...
```

### App Overview

With kpt, you could see an overview of your App.

```shell
$ kpt cfg count sampleapp
ApplicationConfiguration: 1
Component: 1

> TraitDefinition: 1

Reviewer: Definition?

kpt directly use GitHub repo as App Repository, so no action needed.

With the help of kpt, we could directly use GitHub Repo as App Repository without organizing apps in any fixed format.

### Release your application to App Repository

So release your OAM app only needs two steps.

1. Create/Fork a github repo.

2. Commit and push your app.
    ```shell
    git add sampleapp/
    git commit -m "my sampleapp commit"
    git remote add origin git@github.com:<your-account>/<your-app-repo>.git
    git push -u origin master
    ```

### Fetch OAM app from remote Repository

Using our [example repository](https://github.com/oam-dev/sample/tree/master/5.OAM_KPT_Demo/repository/) for this demo.

You could fetch OAM app from remote Repository using [kpt pkg get](https://googlecontainertools.github.io/kpt/reference/pkg/get/).

#### Command

```shell
kpt pkg get https://github.com/oam-dev/samples.git/5.OAM_KPT_Demo/repository/sampleapp sampleapp
```

#### Output

```output
fetching package /5.OAM_KPT_Demo/repository/sampleapp from https://github.com/oam-dev/samples to sampleapp
```

```shell
➜  kpt tree sampleapp
sampleapp
├── Kptfile
├── appconfig.yaml
├── component.yaml

> ├── traitdef.yaml

Reviewer: XxxDefinition should not be part of the app. Please fix this and same issues below.

Author: sorry, fixed

# Managing custom Kubernetes applications with kpt

This guidance will demonstrate how to use kpt to manage custom Kubernetes applications (i.e. apps defined by CRDs instead of raw Kubernetes workloads).

> We will use [Open Application Model](https://github.com/oam-dev/spec)(OAM) to describe the app for standardization and portability consideration. But you are free to use any Custom Resource to describe your application as well.

Reviewer: ```suggestion
We will use [Open Application Model](https://github.com/oam-dev/spec)(OAM) to describe the app for standardization and portability consideration. But you are free to use any Custom Resource to describe your application as well. [Crossplane](https://github.com/crossplane/crossplane) will be installed as the OAM runtime for Kubernetes.
```

Pioneered by [Kubernetes community](https://twitter.com/bgrant0607/status/1221485437153243137), Configuration-as-Data emphasizes that "configuration should be treated as data and leverage pipelines for manipulation and policy enforcement". In Kubernetes, Configuration-as-Data approach builds upon the design of the [Kubernetes resource model (KRM)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/resource-management.md). As a result, today any resource we applied to Kubernetes is a piece of data that represents the desired state for certain part of application or infrastructure in the real system.

With the heart of Configuration-as-Data, a typical Kubernetes native application management workflow just looks like a "pipeline". See the picture below:

![img](./resource/kpt_oam.png)

In this workflow, kpt is the manipulator of data. Stored in data source like Git, the original data (e.g. deployment.yaml) will pass through a pipeline of kpt functionalities to be manipulated into the desire state step by step. For example, `labels` added, `replicas` modified and `image` updated etc. 

> So what is OAM then? OAM is the format of data. More accurately, OAM is a complementary data format to Kubernetes so it can expose higher level abstractions such as "application" to developers.

Reviewer: ```suggestion
So what is OAM then? OAM is the format of data. More accurately, OAM is a specific data format so Kubernetes can expose higher level abstraction such as "application" to developers.
```

In that sense, it's a natural match for using kpt to manipulate the data which is formatted by OAM specification.

## Pre-requisites

> Install OAM plugin to Kubernetes. In this demo, we will use Crossplane by following [this installation guides](https://github.com/oam-dev/crossplane-oam-sample#installation).

Reviewer: ```suggestion
Install OAM Kubernetes runtime (Crossplane) by following [its installation guides](https://github.com/oam-dev/crossplane-oam-sample#installation).
```

    git remote add origin git@github.com:<your-account>/<your-app-repo>.git
    git push -u origin master
    ```

### Fetch OAM app from remote Repository

Using our [example repository](https://github.com/oam-dev/sample/tree/master/5.OAM_KPT_Demo/repository/) for this demo.

You could fetch OAM app from remote Repository using [kpt pkg get](https://googlecontainertools.github.io/kpt/reference/pkg/get/).

#### Command

```shell
kpt pkg get https://github.com/oam-dev/samples.git/5.OAM_KPT_Demo/repository/sampleapp sampleapp
```

#### Output

```output
fetching package /5.OAM_KPT_Demo/repository/sampleapp from https://github.com/oam-dev/samples to sampleapp
```

```shell
➜  kpt tree sampleapp
sampleapp
├── Kptfile
├── appconfig.yaml
├── component.yaml
├── traitdef.yaml
└── workloaddef.yaml

0 directories, 5 files
```

### Install sample app

```shell
$ kubectl apply -f sampleapp/
component.core.oam.dev/example-component created
applicationconfiguration.core.oam.dev/example-appconfig created
traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev created
workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created
```

Several resources are created:

> * Component is the description of your application.

Reviewer: ```suggestion
* Component is the description of what you want to deploy.
```

Check the underlying Deployment instance generated by OAM.

```shell
$ kubectl get deploy
NAME                                    READY   UP-TO-DATE   AVAILABLE   AGE
example-appconfig-workload-deployment   3/3     3            3           114s
```

### Sync with Remote App Repository

When some changes occurred both local and remote apps, you could sync and merge with kpt.

For example, we changed our sampleapp and tag it as `v0.1.0`. Then our local sampleapp also be changed.

```shell
kpt pkg update sampleapp@v0.1.0 --strategy=resource-merge
```

Ref to [update section](https://googlecontainertools.github.io/kpt/guides/consumer/update/#commit-local-changes) of kpt for more details.

### Parameter Setting

[kpt setters](https://googlecontainertools.github.io/kpt/guides/consumer/set/) is a powerful feature which naturally matches to the idea of "separate concerns" design from OAM.

In Open Application Model, developers can claim certain fields in the application YAML as "configurable", so in the following workflow, operators (or the platform) will be allowed to modify these fields.

Now this workflow can be easily achieved with help of kpt.

> OAM parameter is very important feature as it allows App Developer tell their demand to App Operator.

Reviewer: Delete L123-186, as suggested by https://github.com/oam-dev/samples/pull/31#discussion_r406561090, how OAM set parameters is off topic for this blog post.

```shell
$ kpt cfg list-setters sampleapp/
  NAME              VALUE               SET BY                   DESCRIPTION             COUNT
  image           nginx:1.16.1        sampleapp developer   use to set image for component   1
  instance-name   example-component   sampleapp developer   use to set an instance name      1
```

It's very clear and easy to understand.

#### Set Value by App Operator

Then the application operator could set `instance-name` with a new name like this:

```shell
$ kpt cfg set sampleapp/ instance-name test-component
set 1 fields
```

Check the component and you will find the instane name has been changed.

```shell
$ cat sampleapp/component.yaml
apiVersion: core.oam.dev/v1alpha2
kind: Component
metadata:
  name: test-component # {"$ref":"#/definitions/io.k8s.cli.setters.instance-name"}
spec:
  parameters:
  - name: instance-name
    fieldPaths:
    - metadata.name
...
```

### App Overview

With kpt, you could see an overview of your App.

```shell
$ kpt cfg count sampleapp
ApplicationConfiguration: 1
Component: 1
TraitDefinition: 1
WorkloadDefinition: 1
```

> So in the sampleapp, we have one ApplicationConfiguration, Component, TraitDefinition and WorkloadDefinition.

Reviewer: It's better to make XxxDefinitions as part of Crossplane installation, ref: https://github.com/oam-dev/crossplane-oam-sample/issues/1

    git remote add origin git@github.com:<your-account>/<your-app-repo>.git
    git push -u origin master
    ```

### Fetch OAM app from remote Repository

Assuming you have already built an OAM App repository by GitHub.

Using our [example repository](https://github.com/oam-dev/sample/tree/master/5.OAM_KPT_Demo/repository/) for this demo.

You could fetch OAM app from remote Repository using [kpt pkg get](https://googlecontainertools.github.io/kpt/reference/pkg/get/).

#### Command

```
kpt pkg get https://github.com/oam-dev/samples.git/5.OAM_KPT_Demo/repository/sampleapp sampleapp
```

#### Output

```
fetching package /5.OAM_KPT_Demo/repository/sampleapp from https://github.com/oam-dev/samples to sampleapp
```

```
➜  kpt tree sampleapp
sampleapp
├── Kptfile
├── appconfig.yaml
├── component.yaml
├── traitdef.yaml
└── workloaddef.yaml

0 directories, 5 files
```

### Install sample app

```
$ kubectl apply -f sampleapp/
applicationconfiguration.core.oam.dev/example-appconfig created
component.core.oam.dev/example-component created
traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev created
workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev created

> ```

Reviewer: Should explain what are the above resources. e.g.

- `Component` is the description of your application.
- `ApplicationConfiguration` is the description of operational policies for your `Component`.
- `TraitDefinition` and `WorkloadDefinition` are internal CRDs for OAM plugin to use.

              protocol: "TCP"
          env:
            - name: WORDPRESS_DB_PASSWORD
              value: ""
  parameters:
    - name: instance-name
      required: true
      fieldPaths:
        - metadata.name
    - name: image
      fieldPaths:
        - spec.containers[0].image
```

The App Developer left two parameters which are `instance-name` and `image`.

Then in AppConfig, App Operator could overwrite this parameter like below.

```yaml
---
apiVersion: core.oam.dev/v1alpha2
kind: ApplicationConfiguration
metadata:
  name: example-appconfig
spec:
  components:
    - componentName: example-component
      parameterValues:
        - name: instance-name
          value: example-appconfig-workload
        - name: image
          value: nginx:latest
      traits:
        - trait:
            apiVersion: core.oam.dev/v1alpha2
            kind: ManualScalerTrait
            metadata:
              name:  example-appconfig-trait
            spec:
              replicaCount: 3
```

With KPT setters, this feature become more clear and useful. And you don't need to declare parameters in OAM yamls. Let's go through with it.

#### Create setter by App Developer

> As the app developer need to add two parameter for his component, we add two KPT setters here:

Reviewer: ```suggestion
Let's say the developer need to claim two fields as "configurable" for his application, he can add two kpt setters here:
```

```
$ kpt cfg create-setter sampleapp/ instance-name example-component --field "metadata.name" --description "use to set an instance name" --set-by "sampleapp developer"
```

```
$ kpt cfg create-setter sampleapp/ image nginx:1.16.1 --field "spec.workload.spec.containers[0].image" --description "use to set image for component" --set-by "sampleapp developer"
```

Then the app operator could see which parameters are available in this component like below:

```
$ kpt cfg list-setters sampleapp/
  NAME              VALUE               SET BY                   DESCRIPTION             COUNT
  image           nginx:1.16.1        sampleapp developer   use to set image for component   0
  instance-name   example-component   sampleapp developer   use to set an instance name      1
```

It's very clear and easy to understand.

#### Set Value by App Operator

> Then the App Operator could set `instance-name` with a new name like this:

Reviewer: ```suggestion
Then the application operator could set `instance-name` with a new name like this:
```

```
$ kpt live init sampleapp
Initialized: ../sampleapp/grouping-object-template.yaml
```

```
$ kpt live apply sampleapp --wait-for-reconcile
configmap/grouping-object-a4e18a93 created
applicationconfiguration.core.oam.dev/example-appconfig configured
component.core.oam.dev/test-component created
traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev unchanged
workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev unchanged
5 resource(s) applied. 2 created, 2 unchanged, 1 configured
configmap/grouping-object-a4e18a93 is Current: Resource is always ready
applicationconfiguration.core.oam.dev/example-appconfig is Current: Resource is current
traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev is Unknown:
workloaddefinition.core.oam.dev/containerizedworkloads.core.oam.dev is Unknown:
resources failed to the reached Current status
traitdefinition.core.oam.dev/manualscalertraits.core.oam.dev pruned
configmap/grouping-object-a4e18a93 pruned
2 resource(s) pruned
```

Currently `kpt live apply` seems not work well and we have created an [issue #498](https://github.com/GoogleContainerTools/kpt/issues/498).

> ### Plugin and Pipeline

Author: OK
